sensor/common/networkflow/updatecomputer/transition_based.go

```diff
@@ -0,0 +1,509 @@
+// the type of transition.
+func categorizeUpdateNoPast(
+	currTS timestamp.MicroTS,
+	key string, deduper *set.StringSet, mutex *sync.RWMutex) (bool, TransitionType) {
+
+	// Variables for ease of reading
+	isClosed := currTS != timestamp.InfiniteFuture
+
+	// UNKNOWN -> CLOSED
+	if isClosed {
+		// We are unable to check whether this is closed->closed or open->closed.
+		// The latter is assumed to be on the safe side and properly update the deduper.
+		// Update to Central will always be sent, even if that was indeed a closed->closed transition.
+		return true, TransitionTypeOpen2Closed
+	}
+	// UNKNOWN -> OPEN - as last check due to costly search in the deduper
+	seenPreviouslyOpen := concurrency.WithRLock1(mutex, func() bool {
+		return deduper.Contains(key)
+	})
+	if seenPreviouslyOpen {
+		return false, TransitionTypeOpen2Open
+	}
+	return true, TransitionTypeNew2Open
+}
+
+// ComputeUpdatedEndpoints computes endpoint updates to send to Central in the current tick.
+// This method doesn't rely on the state of closed endpoints from the past; each closed endpoint generates an update.
+func (c *TransitionBased) ComputeUpdatedEndpoints(current map[indicator.ContainerEndpoint]timestamp.MicroTS) []*storage.NetworkEndpoint {
+	var updates []*storage.NetworkEndpoint
+	ee := EndpointEnrichedEntity
+	UpdateEventsGauge.Reset()
+	if len(current) == 0 {
+		return c.cachedUpdatesEp
+	}
+
+	// Process currently enriched endpoints one by one, categorize the transition, and generate an update if applicable.
+	for ep, currTS := range current {
+		key := ep.Key(c.hashingAlgo)
+		// Based on the categorization, we calculate the transition and whether an update should be sent.
+		update, transition := categorizeUpdateNoPast(currTS, key, c.deduper[ee], &c.deduperMutex)
+		updateMetrics(update, transition, ee)
+		// Each transition may require updating the deduper.
+		// We cannot update the deduper right away, because Central may be offline, so we must store the operations
+		// to execute on the deduper and execute them only when sending to Central is successful.
+		action := getDeduperAction(transition)
+		switch action {
+		case deduperActionAdd:
+			c.deduper[ee].Add(key)
+		case deduperActionRemove:
+			c.deduper[ee].Remove(key)
+		}
+		if update {
+			updates = append(updates, ep.ToProto(currTS))
+		}
+	}
+	// Store into cache in case sending to Central fails.
+	c.cachedUpdatesEp = slices.Grow(c.cachedUpdatesEp, len(updates))
+	c.cachedUpdatesEp = append(c.cachedUpdatesEp, updates...)
+	// Return concatenated past and current updates.
+	return c.cachedUpdatesEp
+}
+
+// ComputeUpdatedProcesses computes process updates to send to Central in the current tick.
+// This method doesn't rely on the state of closed processes from the past; each closed process generates an update.
+func (c *TransitionBased) ComputeUpdatedProcesses(current map[indicator.ProcessListening]timestamp.MicroTS) []*storage.ProcessListeningOnPortFromSensor {
```

Comment:
[nit] This looks pretty much like `ComputeUpdatedEndpoints` (mod the FF check, the EE, and cache map). I noticed `categorizeUpdateNoPast` is re-used. I wonder if it's worth re-using more parts of this logic. In other words: Do you think it's worth extracting this logic to something like `computeUpdateNoPast` and pass the EE type and the cache map as parameters?

Reply:
I gave it a try - it is not the most beautiful result, but I think it is worth it: [48cb81b](https://github.com/stackrox/stackrox/pull/16522/commits/48cb81b3de52b7745eb3f5b13311d752034f7664)